graph/complement_test.go

```diff
@@ -28,7 +28,7 @@ var complementTests = []struct {
 
 func TestComplement(t *testing.T) {
 	for _, test := range complementTests {
-		n := test.g.Nodes().Len()
+		n := len(graph.NodesOf(test.g.Nodes()))
```

Comment:
So, the `graph.Nodes` interface allows implementations to return -1 to indicate that the iterator doesn't want to spend the effort to calculate the number of nodes that it will iterate through until it actually does the iteration (so actually never). This test wants to find out how many nodes the iterator has and to do this it must actually materialise the nodes.

Reply:
OK. I think the naming (Len) is a little confusing. An inexperienced reader would expect Len() >= 0. How about LazyLen?

Reply:
This reflects the use of -1 length in `[http.Request](https://golang.org/pkg/net/http/#Request).ContentLength` to indicate unknown length. The behaviour is documented and has been there for a while.

We could add a `graph.LengthOf(graph.Iterator) int64` function that counts rather than copies and discards, though it has sort of limited use.

Reply:
Fair enough.